Kernel/ACPI/DMIDecoder.cpp

```diff
@@ -0,0 +1,257 @@
+#include <Kernel/ACPI/DMIDecoder.h>
+#include <Kernel/StdLib.h>
+#include <Kernel/VM/MemoryManager.h>
+
+static DMIDecoder* s_dmi_decoder;
+
+//#define SMBIOS_DEBUG
+
+#define SMBIOS_BASE_SEARCH_ADDR 0xf0000
+#define SMBIOS_END_SEARCH_ADDR 0xfffff
+#define SMBIOS_SEARCH_AREA_SIZE (SMBIOS_END_SEARCH_ADDR - SMBIOS_BASE_SEARCH_ADDR)
+
+DMIDecoder& DMIDecoder::the()
+{
+    if (s_dmi_decoder == nullptr) {
+        s_dmi_decoder = new DMIDecoder(true);
+    }
+    return *s_dmi_decoder;
+}
+
+void DMIDecoder::initialize()
+{
+    if (s_dmi_decoder == nullptr) {
+        s_dmi_decoder = new DMIDecoder(true);
+    }
+}
+
+void DMIDecoder::initialize_untrusted()
+{
+    if (s_dmi_decoder == nullptr) {
+        s_dmi_decoder = new DMIDecoder(false);
+    }
+}
+
+void DMIDecoder::initialize_parser()
+{
+    if (m_entry32bit_point != nullptr || m_entry64bit_point != nullptr) {
+        m_operable = true;
+        kprintf("DMI Decoder is enabled\n");
+        if (m_entry64bit_point != nullptr) {
+            kprintf("DMIDecoder: SMBIOS 64bit Entry point @ P 0x%x\n", m_entry64bit_point.ptr());
+            m_use_64bit_entry = true;
+            m_structure_table = OwnPtr<SMBIOS::TableHeader>((SMBIOS::TableHeader*)m_entry64bit_point->table_ptr);
+            m_structures_count = m_entry64bit_point->table_maximum_size;
+            m_table_length = m_entry64bit_point->table_maximum_size;
+        } else if (m_entry32bit_point != nullptr) {
+            kprintf("DMIDecoder: SMBIOS 32bit Entry point @ P 0x%x\n", m_entry32bit_point.ptr());
+            m_use_64bit_entry = false;
+            m_structure_table = OwnPtr<SMBIOS::TableHeader>((SMBIOS::TableHeader*)m_entry32bit_point->legacy_structure.smbios_table_ptr);
+            m_structures_count = m_entry32bit_point->legacy_structure.smbios_tables_count;
+            m_table_length = m_entry32bit_point->legacy_structure.smboios_table_length;
+        }
+        kprintf("DMIDecoder: Data table @ P 0x%x\n", m_structure_table.ptr());
+        enumerate_smbios_tables();
+    } else {
+        m_operable = false;
+        kprintf("DMI Decoder is disabled. Cannot find SMBIOS tables.\n");
+    }
+}
+
+void DMIDecoder::enumerate_smbios_tables()
+{
+    u32 table_length = m_table_length;
+    SMBIOS::TableHeader* physical_table = m_structure_table.ptr();
+
+    auto region = MM.allocate_kernel_region(PAGE_ROUND_UP(table_length), "DMI Decoder Entry Point Finding", Region::Access::Read);
```

Comment:
If we want this region (and all the other `Region` objects in these patches) to be guaranteed to stay alive, and for the address space to not be stolen by other things later on, we have to keep the `Region` objects somewhere. Otherwise the `Region` destructor will deallocate the virtual addresses used.

The most obvious thing would be to store them in `RefPtr<Region>` class members.

Reply:
The whole point is to destruct those regions since I don't care about the SMBIOS entry point.
Even worse, there is a 64 bit entry point version that some BIOSes can implement. So, in order to abstract access to SMBIOS data, I only care about the values (and not the mapping), and therefore I only parse it one time and forget about it.
Other regions that I actually want to keep I put as class members, so they won't go away as soon as I go out of the method.

Reply:
Ah fair enough, I misunderstood that these were not needed to persist :)